pkg/kubectl/apis/manifest/v1alpha1/types.go

```diff
@@ -0,0 +1,140 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Package contains all the metadata of the package and drives package searching and browsing,
+// and support the fork/rebase upgrade workflow. It can be used by something like an app registry.
+type Package struct {
+	metav1.TypeMeta `json:",inline" yaml:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// Name of an instance of the package.
+	// There may be multiple instances with the same package and different packageVersions.
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// Description of the package.
+	Description string `json:"description,omitempty" yaml:"description,omitempty"`
+
+	// An pointer to the icon.
+	Icon string `json:"icon,omitempty" yaml:"icon,omitempty"`
+
+	// Search keywords for the package.
+	Keywords []string `json:"keywords,omitempty" yaml:"keywords,omitempty"`
+
+	// Homepage of the application package.
+	Home string `json:"home,omitempty" yaml:"home,omitempty"`
+
+	// Source specifies the upstream URL, e.g. https://github.com/foo/bar.git, file://host/path, etc.
+	// hosting the resource files specified in Base and Overlays.
+	// This is useful in the fork/rebase workflow.
+	Source string `json:"source,omitempty" yaml:"source,omitempty"`
+
+	// Version of the package.
+	PackageVersion string `json:"packageVersion,omitempty" yaml:"packageVersion,omitempty"`
+}
+
+// Manifest has all the information to expand of generate the k8s api resources.
+// It can be used by kubectl or some other tooling.
+// A manifest could be either a Base or an Overlay.
+type Manifest struct {
+	metav1.TypeMeta `json:",inline" yaml:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// Name of an instance of the manifest.
+	// There may be multiple instances with the same manifest and different manifestVersions.
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// Description of the manifest.
+	Description string `json:"description,omitempty" yaml:"description,omitempty"`
+
+	// Version of the manifest.
+	ManifestVersion string `json:"manifestVersion,omitempty" yaml:"manifestVersion,omitempty"`
+
+	// Partial name that will prefix the name of the base resources.
```

Comment:
It should prefix the names of all resources mentioned (e.g., also configmaps and secrets generated).

Reply:
>  also configmaps and secrets generated

I feel a little confused here due to double name prefix of configmap and secret.
What the final name of a configmap generated will look like? like `{mainfest.namePrefix}-{manifest.configmaps[0].namePrefix}-{hash(content of configmap)}`?

Reply:
Maybe the configmap and secret namePrefix should just be called them.

The Manifest's namePrefix should prefix everything. It may be something like `myapp-`. I'd expect it to be in overlay manifests only, because the base package can just name resources however it likes.

The secret/configmap name prefix is logically the name of the resource to generate, such as tlscert, but will be suffixed by the hash of the contents. I expect that name to be specified wherever the secrets/configmaps are, which will sometimes be in the base and sometimes in overlays, in the case that the overlays add additional resources to the base. For example, I wouldn't expect real secrets to be in a package repo, but they could be added in a fork of the package, or in overlays in the case of multiple instances of the package.